src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,324 @@
+          await mkdir(dataPath, { recursive: true });
+          break;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  return instance;
+}
+
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  let dirEntries: Dirent[];
+  let processes: ProcessDescriptor[];
+  let someInstancesFailed = false;
+
+  try {
+    [dirEntries, processes] = await Promise.all([
+      readdir(dataPath, { withFileTypes: true }),
+      psList()
+    ]);
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
+      throw err;
+    }
+    // instances directory does not (yet) exist, so there cannot be any instances
+    return [];
+  }
+  const instances: DetachedInstance[] = [];
+
+  const loadingInstancesInfos = dirEntries.map(async dirEntry => {
+    const filename = dirEntry.name;
+    const { name: instanceName, ext } = path.parse(filename);
+
+    let failureReason: string;
+
+    if (ext !== ".json") {
+      failureReason = `"${filename}" does not have a .json extension`;
+    } else {
+      let instance: DetachedInstance;
+      try {
+        // getDetachedInstanceByName() throws if the instance file is not found or
+        // cannot be parsed
+        instance = await getDetachedInstanceByName(instanceName);
+      } catch (err: any) {
+        failureReason = err.message;
+      }
+      if (instance) {
+        const matchingProcess = processes.find(p => p.pid === instance.pid);
+        if (!matchingProcess) {
+          failureReason = `Process with PID ${instance.pid} could not be found`;
+        } else if (matchingProcess.cmd !== instance.cmd) {
+          failureReason = `Process with PID ${instance.pid} does not match ${instanceName}`;
+        } else {
+          instances.push(instance);
+        }
+      }
+    }
+
+    if (failureReason !== undefined) {
+      someInstancesFailed = true;
+      const fullPath = path.join(dataPath, filename);
+      let resolution: string;
+      if (dirEntry.isDirectory()) {
+        const reason = `"${filename}" is a directory`;
+        try {
+          await rm(fullPath, { recursive: true });
+          failureReason = reason;
+        } catch {
+          resolution = `"${filename}" could not be removed`;
+        }
+      } else {
+        try {
+          await unlink(fullPath);
+        } catch {
+          resolution = `"${filename}" could not be removed`;
+        }
+      }
+
+      console.warn(
+        `Failed to load instance data. ${failureReason}. ${
+          resolution || `"${filename}" has been removed`
+        }.`
+      );
+    }
+  });
+
+  await Promise.all(loadingInstancesInfos);
+
+  if (someInstancesFailed) {
+    console.warn(
```

Comment:
I do this at the end, because if there's multiple errors, we only want to say this once.